### PR TITLE
Attempt to add Geometry datatype support for PostgreSQL (PostGis)

### PR DESCRIPTION
**Pull request by mibrito**
As discussed on https://github.com/sequelize/sequelize/issues/2839, it would be nice to have this feature on sequelize. Particulary I'm working in a project using sequelize that needs to be able to insert and retrieve geographic data from PostgreSQL with <a href="http://postgis.net/">PostGIS</a>. So I made a fork of the sequelize and tried to insert some initial steps to achive this functionality.

Basicaly was created a new datatype GEOMETRY and some of the functions were adapted to store geometries using GeoJSON objects (<a href="http://www.postgis.org/docs/ST_AsGeoJSON.html">ST_GeomFromGeoJSON</a>) and retrieve them as GeoJSON (<a href="http://postgis.org/docs/ST_GeomFromGeoJSON.html">ST_AsGeoJSON</a>).

The new datatype created is based on <a href="http://postgis.net/docs/AddGeometryColumn.html">Geometry Type Definition</a> syntax added on PostGIS version 2.0.0:

<code>Sequelize.GEOMETRY(geometry_type::asString)</code>
Allows the method .sync() to create a new column of the type Geometry(geometry_type), which by default on PostGIS has srid equals -1.

<code>Sequelize.GEOMETRY(geometry_type::asString, srid::asNumber)</code>
Allows the method .sync() to create a column of type Geometry(geometry_type, srid).

<code>postgres.QueryGenerator.escape()</code>
Was modified to "escape" geometry data using PostGIS function ST_GeomFromGeoJSON.

<code>abstract.QueryGenerator.selectQuery()</code>
Was modified to retrieve data as GeoJSON using PostGIS function ST_AsGeoJSON.

<code>Instance.prototype.set()</code>
Was modified to parse GeoJSON object as JSON.

Model now has a vector called <code>_geometryAttributes</code>, a function called <code>_hasGeometryAttributes</code> and a function called <code>_isGeometryAttribute</code>.

As a first attempt it still needs some tests and adjustments, so i'm sharing it now to have some feedback and ideas on the changes that are needed.

### CODE REVIEW

File: lib/dialects/postgres/data-types.js [hunk 1]
```diff
@@ -80,6 +80,11 @@ BLOB.prototype.toSql = function() {
   return 'BYTEA';
 };
 
+var GEOMETRY = function() {
```
Comment (janmeier) [inline lib/dialects/postgres/data-types.js]: There is no reason to have any special code in postgres - having all the code in abstract datatypes is fine

Comment (janmeier) [inline lib/dialects/postgres/data-types.js]: Dialect specific datatypes are only used if a data type is only supported by a single dialect, or if a dialect uses a different type (for example postgres blobs = bytea)

The geometry type is supported by mysql and mssql as well

File: lib/dialects/postgres/query-generator.js [hunk 2]
```diff
@@ -893,6 +893,8 @@ module.exports = (function() {
         return 'ARRAY[' + value.map(function (v) {
           return SqlString.escape(JSON.stringify(v), false, this.options.timezone, this.dialect, field);
         }, this).join(',') + ']::JSON[]';
+      } else if (field && (field.type instanceof DataTypes.GEOMETRY || field.type instanceof DataTypes.GEOMETRY)) {
```
Comment (janmeier) [inline lib/dialects/postgres/query-generator.js]: Did you mean `or field.type instanceof Array(Geometry)`?

File: test/unit/sql/data-types.test.js [hunk 3]
```diff
@@ -282,6 +282,33 @@ suite(Support.getTestDialectTeaser('SQL'), function() {
...
 
+    suite('GEOMETRY', function () {
+      testsql('GEOMETRY', DataTypes.GEOMETRY, {
+        default: 'GEOMETRY',
+        postgres: 'GEOMETRY'
+      });
+	
+      testsql('GEOMETRY(\'POINT\')', DataTypes.GEOMETRY('POINT'), {
+        default: 'GEOMETRY(POINT)',
+        postgres: 'GEOMETRY(POINT)'
```
Comment (janmeier) [inline test/unit/sql/data-types.test.js]: No reason to have a special case for postgres here when the expected result is the same

**Comment by janmeier** [comment]
Looking good so far! :)

To be able to test this you need to enable postgis in the travis config. http://docs.travis-ci.com/user/using-postgresql/#Using-PostGIS

It seems like geojson is also supported in mysql 5.7 (http://dev.mysql.com/doc/refman/5.7/en/spatial-geojson-functions.html#function_st-geomfromgeojson) - but its still in development, so...
The ideal solution from my point of view would be to use geojson as the abstraction for all dialects in sequelize. For example, inserting `{ "type": "Point", "coordinates": [125.6, 10.1] }` in sequelize would translate to `GeomFromText('POINT(125.6, 10.1)')` in mysql https://dev.mysql.com/doc/refman/5.0/en/populating-spatial-columns.html.

When fetching data again, the data is returned in the well-known text format, but it should be as simple as pluging in https://www.npmjs.com/package/wellknown or similar to return geojson

Still needs unit tests for escaping geo types, and integration tests, but you probably already know that :)


It could be super-duper awesome if you can get this to work for mysql and mssql as well using WKT - it shouldn't be a lot of added work.

**Comment by janmeier** [comment]
@mibrito , @mikeringrose, @Ghirro Which of you are still working on this feature, and whats the best way to move forward with it? Are any of you tempted by the posibiltiy of implenting this using wkt in mysql / mssql and geojson in PG. If we can only get the postgres support in for now that'd also be fine, but I'd love to support this cross-dialect

**Comment by mikeringrose** [comment]
@janmeier since i'll be using it in production shortly. i plan on writing more test and add mysql support this weekend.

**Comment by janmeier** [comment]
@mikeringrose Looking forward to it :)

**Comment by arneke** [comment]
Hi, very cool stuff mibrito, I was about to start hacking my own support into the escape functions before I saw this pull request.

I've found one issue while testing this on legacy tables, if the column name is not equal to the attribute name, i.e. 
```
  the_geom: {
    type: Sequelize.GEOMETRY,
    field: 'the_geom'
  }
```

Then the abstract query generator fails becase in "mainModel._isGeometryAttribute(attr)" , the value of attr is actually "the_geom,geometry"

(The pull request appears to be for 2.1 instead of 3.0?)

File: lib/dialects/abstract/query-generator.js [hunk 4]
```diff
@@ -901,6 +901,8 @@ module.exports = (function() {
             }
           }
           attr = [attr[0], self.quoteIdentifier(attr[1])].join(' AS ');
+        } else if (mainModel && mainModel._isGeometryAttribute(attr)){
+          attr = 'ST_AsGeoJSON('+self.quoteIdentifiers(attr)+') AS '+ self.quoteIdentifiers(attr);
```
Comment (arneke) [inline lib/dialects/abstract/query-generator.js]: attr may be composed of both the column name and the alias, comma separated. Then the lookup doesn't work.

**Comment by mshahriarinia** [comment]
Any updates on this? This seems to be a valid feature that is not there. yet (?)

**Comment by janmeier** [comment]
Closed in favor of https://github.com/sequelize/sequelize/pull/3828

**Pull request closed without merging**